### PR TITLE
Fix bad shop group condition and delete old delivery records

### PR DESCRIPTION
**Pull request by matks**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Fix bad shop group condition and delete old delivery records. Port of https://github.com/PrestaShop/PrestaShop/pull/8816
| Type?         | bug fix
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | Partially https://github.com/PrestaShop/PrestaShop/issues/9592. See last comment. 
| How to test?  | 😅 

So this PR does 2 things:

- it fixes a bad SQL condition where it's obvious id_shop should be id_shop_group
- it deletes old delivery SQL table records when new delivery records are added for the new carrier
What do I mean by "new carrier" ? Actually when carrier is edited, it is, a new carrier is created with new informations and an old record, unmodified, stays hidden in database.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/15615)

### CODE REVIEW

File: classes/Carrier.php [hunk 1]
```diff
@@ -959,8 +979,9 @@ public function addDeliveryPrice($price_list, $delete = false)
                 Db::getInstance()->execute(
                     'DELETE FROM `' . _DB_PREFIX_ . 'delivery`
                     WHERE ' . (null === $values['id_shop'] ? 'ISNULL(`id_shop`) ' : 'id_shop = ' . (int) $values['id_shop']) . '
-                    AND ' . (null === $values['id_shop_group'] ? 'ISNULL(`id_shop`) ' : 'id_shop_group=' . (int) $values['id_shop_group']) . '
-                    AND id_carrier=' . (int) $values['id_carrier'] .
+                    AND ' . (null === $values['id_shop_group'] ? 'ISNULL(`id_shop_group`) ' : 'id_shop_group=' . (int) $values['id_shop_group']) .
+                    // delete delivery records related to old carrier records
+                    'AND id_carrier IN (' . implode('","', $old_carrier_record_ids) .
```
Comment (matthieu-rolland) [inline classes/Carrier.php:984]: ` 'AND id_carrier IN (` 
I can't find the closing parenthesis (maybe I missed it?)

File: classes/Carrier.php [hunk 2]
```diff
@@ -946,6 +947,25 @@ public function addDeliveryPrice($price_list, $delete = false)
             $keys[] = 'id_shop_group';
         }
 
+        $old_carrier_record_ids = array();
```
Comment (PierreRambaud) [inline classes/Carrier.php:950]:
```suggestion
        $old_carrier_record_ids = [];
```

File: classes/Carrier.php [hunk 3]
```diff
@@ -946,6 +947,25 @@ public function addDeliveryPrice($price_list, $delete = false)
...
+            // The old, unmodified record stays hidden in database.
+
+            $old_carrier_records = Db::getInstance()->executeS('
+                SELECT `id_carrier` 
+                FROM `' . _DB_PREFIX_ . 'carrier` 
+                WHERE `id_reference` = ' . (int) $this->id_reference
+            );
+            if (is_array($old_carrier_records) && count($old_carrier_records) > 0) {
+                $old_carrier_record_ids = array_map(
+                    function ($record) {return $record['id_carrier']; },
```
Comment (PierreRambaud) [inline classes/Carrier.php:963]:
```suggestion
                    function ($record) { return $record['id_carrier']; },
```

File: classes/Carrier.php (hunk 2 above)
Comment (S4hosting) [reply classes/Carrier.php:950]: which one have to use array() or [] ?

Comment (matks) [reply classes/Carrier.php:950]: `[]` is the good and modern way 👍

**Pull request closed without merging**